File: problems/16135 OBB/sol.js
let input = require('fs').
    readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt").
    toString().trim().split("\n");

input = input.map(v => v.trim().split(' ').map(Number));

[x1, y1, x2, y2, x3, y3, x4, y4] = input[0];
[x21, y21, x22, y22, x23, y23, x24, y24] = input[1];


function subtract({x, y}) {
    return {"x": this.x-x, "y":this.y-y};
}
function add({x, y}) {
    return {"x":this.x+x, "y":this.y+y};
}
function dot({x, y}) {
    return Math.abs(this.x*x+this.y*y);
}

let p1 = {"x": x1, "y": y1};
let p2 = {"x": x2, "y": y2};
let p3 = {"x": x3, "y": y3};

let p21 = {"x": x21, "y": y21};
let p22 = {"x": x22, "y": y22};
let p23 = {"x": x23, "y": y23};

let c1 = add.call(p1, p3);
let c2 = add.call(p21, p23);

let a1 = subtract.call(p1,p2);
let a2 = subtract.call(p2, p3);

let b1 = subtract.call(p21,p22);
let b2 = subtract.call(p22, p23);

let d = subtract.call(c1, c2);
function check(u) {
    return dot.call(d, u) >= dot.call(a1, u) + dot.call(a2, u) + dot.call(b1, u) + dot.call(b2, u);
}

u = subtract.call(p1, p2);
let ret = 0;
ret |= check(u);
u = subtract.call(p2, p3);
ret |= check(u);

u = subtract.call(p21, p22);
ret |= check(u);
u = subtract.call(p22, p23);
ret |= check(u);

console.log(+!ret);


// 	9344KB	92ms

File: problems/백준/16135 OBB/sol.cpp
#include<set>
#include<iostream>
#include <set>

#define pii pair<int, int>

using namespace std;

struct Point{
    int x, y;
    Point(int _x, int _y): x(_x), y(_y) {}
    Point subtract(Point p) {
        return Point(this->x-p.x, this->y-p.y);
    }
    Point add(Point p) {
        return Point(this->x+p.x, this->y+p.y);
    }
    int dot(Point p) {
        int a = this->x*p.x;
        int b = this->y*p.y;
        return abs(a+b);
    }

};
void _main(int tnum) {
    int x1, y1, x2, y2, x3, y3, x4, y4;
    cin >> x1 >> y1 >> x2 >> y2 >> x3 >> y3 >> x4 >> y4;
    set<pii> st = {{x1, y1}, {x2, y2}, {x3, y3}, {x4, y4}};
    Point p1(x1, y1);
    Point p2(x2, y2);
    Point p3(x3, y3);
    Point p4(x4, y4);

    cin >> x1 >> y1 >> x2 >> y2 >> x3 >> y3 >> x4 >> y4;
    Point p21(x1, y1);
    Point p22(x2, y2);
    Point p23(x3, y3);
    Point p24(x4, y4);
    
    
    set<pii> st2 = {{x1, y1}, {x2, y2}, {x3, y3}, {x4, y4}};

    Point a1 = p1.subtract(p2);
    Point a2 = p2.subtract(p3);
    Point b1 = p21.subtract(p22);
    Point b2 = p22.subtract(p23);
    Point c1 = p1.add(p3);
    Point c2 = p21.add(p23);

    Point d = c1.subtract(c2);
    Point u = p1.subtract(p2);

    auto check = [&]() {
        return d.dot(u) >= a1.dot(u) + a2.dot(u) + b1.dot(u) + b2.dot(u);
    };
    int ret = check();
    u = p2.subtract(p3);
    ret |= check();
    u = p21.subtract(p22);
    ret |= check();
    u = p22.subtract(p23);
    ret |= check();
    cout << !ret << endl;
}

int main() {
    _main(0);
    return 0;
}

//2028KB	0ms